### PR TITLE
Erc20 implementation workaround

### PR DESCRIPTION
**Pull request by davidbancor**
These updates give support for token contracts which don't have a return value for `transfer`, `transferFrom`, and `approve`.

### CODE REVIEW

File: solidity/contracts/BancorNetwork.sol [hunk 1]
```diff
@@ -10,8 +10,10 @@ import './utility/SafeMath.sol';
 import './utility/interfaces/IContractRegistry.sol';
 import './utility/interfaces/IContractFeatures.sol';
 import './utility/interfaces/IWhitelist.sol';
+import './utility/interfaces/ITokenWhitelist.sol';
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol]: what's the logic behind that name? why whitelist?

File: solidity/contracts/BancorNetwork.sol [hunk 2]
```diff
@@ -627,7 +629,56 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractIds, FeatureIds {
...
+
+        if (!tokenWhitelist.whitelistedTokens(_token)) {
+            // if the token isn't whitelisted, we assert on transfer
+            assert(_token.transfer(_to, _amount));
+        } else {
+            uint256 prevBalance = _token.balanceOf(_to);
+            // we have to cast the token contract in an interface which has no return value
+            INonStandardERC20(_token).transfer(_to, _amount);
+            uint256 postBalance = _token.balanceOf(_to);
+            assert(postBalance.sub(prevBalance) == _amount);
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol]: that might not work for tokens that have fees, maybe just check that the new amount is greater than the prev amount?

File: solidity/contracts/BancorNetwork.sol [hunk 3]
```diff
@@ -640,10 +691,10 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractIds, FeatureIds {
 
         // if the allowance is nonzero, must reset it to 0 first
         if (_token.allowance(this, _spender) != 0)
-            assert(_token.approve(_spender, 0));
+            INonStandardERC20(_token).approve(_spender, 0);
```
Comment (yudilevi) [inline solidity/contracts/BancorNetwork.sol]: so worst case the tx will fail?

File: solidity/contracts/converter/BancorConverter.sol [hunk 4]
```diff
@@ -824,6 +826,54 @@ contract BancorConverter is IBancorConverter, SmartTokenController, Managed, Con
...
 
+    /**
+        @dev ensures transfer of tokens, taking into account that some ERC-20 implementations don't return
+        true on success but revert instead
+
+        @param _token     the token to transfer
+        @param _to        the address to transfer the tokens to
+        @param _amount    the amount to transfer
+    */
+    function ensureTransfer(IERC20Token _token, address _to, uint256 _amount) private {
```
Comment (yudilevi) [inline solidity/contracts/converter/BancorConverter.sol]: how about moving these functions to a shared utility?

File: solidity/contracts/helpers/BadERC20Token.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
+pragma solidity ^0.4.24;
+import './interfaces/IBadERC20Token.sol';
```
Comment (yudilevi) [inline solidity/contracts/helpers/BadERC20Token.sol]: Bad or NonStandard?

File: solidity/contracts/utility/TokenWhitelist.sol [hunk 6]
```diff
@@ -0,0 +1,24 @@
+pragma solidity ^0.4.24;
+import './Owned.sol';
+import './interfaces/ITokenWhitelist.sol';
+
+/*
+    Token Whitelist
+
+    manages tokens who don't return true/false on transfer/transferFrom but revert on failure instead 
+*/
+contract TokenWhitelist is ITokenWhitelist, Owned {
```
Comment (yudilevi) [inline solidity/contracts/utility/TokenWhitelist.sol]: maybe just use the IWhitelist contract?

File: solidity/contracts/helpers/BadERC20Token.sol (hunk 5 above)
Comment (davidbancor) [reply solidity/contracts/helpers/BadERC20Token.sol]: "bad" for tests, "nonstandard interface" for production, silly naming

File: solidity/contracts/converter/BancorConverter.sol (hunk 4 above)
Comment (davidbancor) [reply solidity/contracts/converter/BancorConverter.sol]: definitely seems like the right idea, probably will move them to Utils

File: solidity/contracts/BancorNetwork.sol (hunk 3 above)
Comment (davidbancor) [reply solidity/contracts/BancorNetwork.sol]: when calling transferFrom afterwards, if the approve didn't succeed the tx should revert

File: solidity/contracts/BancorNetwork.sol (hunk 2 above)
Comment (davidbancor) [reply solidity/contracts/BancorNetwork.sol]: hmmm, didn't consider that case... we should discuss this further

File: solidity/contracts/BancorNetwork.sol (hunk 1 above)
Comment (davidbancor) [reply solidity/contracts/BancorNetwork.sol]: Not the most intuitive name, just didn't want to spend too long thinking about the name... we should think of a more explanatory name

**Pull request closed without merging**